### PR TITLE
Github Actions workflow for building the toolchain and the SDK

### PR DESCRIPTION
**Pull request by fridtjof**
My original motivation for this was to let you provide "nightly" builds of the SDK and toolchain, so our CI would not have to duplicate effort, but it seems that GitHub Actions does not let you easily download artifacts from other repositories (yet?).

Still, the effort was made, so here's a free CI workflow for your project! :)

On runtimes:

- The toolchain takes roughly 40 minutes to build. Caching is implemented, and a rebuild is triggered whenever the `build-toolchain.sh` script changes (because it contains the gcc, binutils, etc version numbers used to build the toolchain). Unfortunately, jobs can't access past run's artifacts, so that has to uploaded on every run, adding ~2 min of runtime :/ A possible (but less clean) solution would be to merge the two jobs into one.
- Caching does not make sense for the SDK and other builds in this repo, because they obviously change all the time.

(note that this PR does not activate CI yet, for security reasons)

### CODE REVIEW

**Comment by fridtjof** [comment]
Actually, there seems to be a problem with file permissions. I will fix this later today.

**Comment by fridtjof** [comment]
Testing on the fork has concluded, this workflow is now **green** and **ready for review**

**Comment by fridtjof** [comment]
I think I'll add macOS CI in another PR after this one.

Definitely feel free to squash all these commits, they turned out to be more than I anticipated.

**Comment by fridtjof** [comment]
> Might as well add `ubuntu-20.04` too.

Good point, will do!

**Comment by fridtjof** [comment]
It's included now, CI is testing here: https://github.com/shinyblink/Ndless/runs/945763481?check_suite_focus=true
Currently going through a full toolchain build for all three platforms (in parallel), should be done in ~45min

**Comment by fridtjof** [comment]
Should be ready for review now. Really this time.

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:16]: Matrix runs for a job happen in parallel, so there isn't any meaningful difference in execution time. Also, while ancient, 16.04 _is_ still supported ;)

> It's sad that there are only Ubuntu environments available...

It is, I wish they would at least have something like CentOS too...

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:21]: Sure, the reason I did it this way was to shave off a little time, because even if there's a cache hit in the end, the checkout always runs because `build-toolchain.sh` becomes part of the cache key. I can match the checkout used for `build-sdk` though.

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:32]: I'll add it, but it seemed to build fine without it? (or is it an optional dependency?)

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:72]: At least libmpfr-dev seems to be required: https://github.com/shinyblink/Ndless/runs/945668803?check_suite_focus=true

If you can't see the detailed logs, here's the relevant excerpt from `build-sdk` on ubuntu-16.04:

```
make -C ndless-sdk
make[1]: Entering directory '/home/runner/work/Ndless/Ndless/ndless-sdk'
make -C libsyscalls
make[2]: Entering directory '/home/runner/work/Ndless/Ndless/ndless-sdk/libsyscalls'
arm-none-eabi-g++ -mcpu=arm926ej-s -std=c++11 -nostdlib -O3 -fPIE -fno-exceptions -fno-rtti -mlong-calls -Wall -Werror -I ../include/ -I ../thirdparty/nspire-io/include/ -D_TINSPIRE -ffunction-sections -fdata-sections -c stubs.cpp -o stubs.o
/home/runner/work/Ndless/Ndless/ndless-sdk/toolchain/install/libexec/gcc/arm-none-eabi/10.1.0/cc1plus: error while loading shared libraries: libmpfr.so.6: cannot open shared object file: No such file or directory
Makefile:18: recipe for target 'stubs.o' failed
make[2]: *** [stubs.o] Error 1
make[2]: Leaving directory '/home/runner/work/Ndless/Ndless/ndless-sdk/libsyscalls'
make[1]: *** [build-libsyscalls] Error 2
Makefile:14: recipe for target 'build-libsyscalls' failed
make: *** [build-ndless-sdk] Error 2
make[1]: Leaving directory '/home/runner/work/Ndless/Ndless/ndless-sdk'
Makefile:19: recipe for target 'build-ndless-sdk' failed
##[error]Process completed with exit code 2.
```

The non-dev packages might also work for this, but they're different for 16.04 (libmpfr4) and 18.04/20.04 (libmpfr6): https://packages.ubuntu.com/search?keywords=libmpfr

Comment (Vogtinator) [reply .github/workflows/main.yml:72]: That's because the toolchain binaries were built against libmpfr, but then copied into a fresh environment without installing deps. This is one of the reasons why I think that having this in a single job is cleaner.

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:41]: Ideally (if accessing previous artifacts were possible) we could skip uploading the toolchain if it wasn't rebuilt.

In the end I agree (doing it in one go is what we do with github.com/shinyblink/sled anyway) that it's the better solution with regards to runtime, simply because of the artifact overhead.

I'll merge the two jobs, and remove the download part. While uploading the toolchain might be beneficial for developers on CI supported OSs, downloading it again in the same job is unnessecary: the cache mechanism takes care of that much faster.

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:52]: I don't think that's possible. Another argument for merging the jobs :D

**Comment by fridtjof** [comment]
> I'm curious, what are those security reasons?

Runners have access to tokens that allow you to do various privileged things to the repository.
To avoid me (or anyone else) from just extracting them with a malicious workflow submitted through a PR, CI does not run on PRs that originate from forks.

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:16]: From https://github.com/actions/virtual-environments#available-environments:

> Looking for other Linux distributions? We do not plan to offer other Linux distributions. We recommend using Docker if you'd like to build using other distributions with the hosted virtual environments. Alternatively, you can leverage self-hosted runners and fully customize your environment to your needs.

I _guess_ we could use Docker for other distros. Last time I tried, I could not get the Dockerfiles to work though, so I think there is some more work required for that?

**Comment by fridtjof** [comment]
Now building: https://github.com/shinyblink/Ndless/actions/runs/196238815

This includes another full toolchain rebuild, because the cache paths have changed (now also includes the toolchain install directory so we can use it directly).

**Comment by adriweb** [comment]
> > I'm curious, what are those security reasons?
> 
> Runners have access to tokens that allow you to do various privileged things to the repository.
> To avoid me (or anyone else) from just extracting them with a malicious workflow submitted through a PR, CI does not run on PRs that originate from forks.

Note that this got improved 2 days ago :) https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml]: Actually, does this make sense? The virtual machines that the hosted runners are in only have 2 vCPUs:

https://docs.microsoft.com/de-de/azure/virtual-machines/dv2-dsv2-series#dsv2-series

Comment (Vogtinator) [reply .github/workflows/main.yml]: The rule of thumb is generally 2*nCPU jobs, as long as there is enough RAM, which there should be.

**Comment by Vogtinator** [comment]
> > > I'm curious, what are those security reasons?
> > 
> > Runners have access to tokens that allow you to do various privileged things to the repository.
> > To avoid me (or anyone else) from just extracting them with a malicious workflow submitted through a PR, CI does not run on PRs that originate from forks.
> 
> Note that this got improved 2 days ago :) https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/

I'm pretty sure CIs other than GitHub's NIH had that figured out from the start already... That blog post doesn't tell how it's looking for free public organizations and repos and on github.com, maybe that's still broken?

File: .github/workflows/main.yml
Comment (Vogtinator) [inline .github/workflows/main.yml:36]: Does this save the toolchain twice now?

File: .github/workflows/main.yml
Comment (Vogtinator) [reply .github/workflows/main.yml:16]: > Matrix runs for a job happen in parallel, so there isn't any meaningful difference in execution time.

Well, the slowest one "wins", so having one job is still faster than three in parallel. Also makes false negatives more likely.

> Also, while ancient, 16.04 is still supported ;)

So are ancient versions of RHEL and SLES, which nobody ever wants to deal with ;-)

I don't mind `ubuntu-20.04` so much, but `ubuntu-16.04` should IMO go.

> I _guess_ we could use Docker for other distros. Last time I tried, I could not get the Dockerfiles to work though, so I think there is some more work required for that?

Getting `Dockerfile`s into the mix would just make it slower, more fragile and rely on more outside sources, so I'm not a fan of that either... For now this should be enough as I'm testing using a non-Debian distro myself locally.

macOS support would make quite a difference though, but with the downside that it'd be moved from a "PRs welcome" level of support to a "build succeeds" one, but I do not have macOS running anywhere so I would not be able to troubleshoot any further than the build log.

If Windows + WSL supports different distros, that would definitely be nice as that would cover two options at once.

**Comment by adriweb** [comment]
Well I could take care of macOS maintenance, but first we have to take care of the other issue/pr so that it can be enabled here on CI :)

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:36]: Yes, once for further use, and once for the artifact uploaded later. I suppose we could just offer the install directory as an artifact instead of making a .tar.gz

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:16]: Okay, I'll remove 16.04 then.

I guess I'll offer help with macOS on a when-I-have-time basis, if that's any relief. I'll also gladly implement a macOS job later, once this is merged. I'm planning to reuse that for `sled` anyway.

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml]: Actually, we should see if the day-to-day runtime changes when we're uploading a directory as an artifact, instead of a cached .tar.gz.
A job is running right now, but it's rebuilding the toolchain (because the paths changed): https://github.com/shinyblink/Ndless/actions/runs/196901972
I will trigger another run later, to get a runtime with the new artifact upload.

If the runtime regresses meaningfully, I would revert to how it was before - the tar.gz doesn't hurt anything, and only gets made when the toolchain is built anyway.

Comment (fridtjof) [reply .github/workflows/main.yml]: The results are in:

With uploading a cached `toolchain.tar.gz` (~400-450MB):
https://github.com/shinyblink/Ndless/actions/runs/196901382

=> about 3:45 minutes

With uploading the cached toolchain install directory as-is (~1-1.5 GB):
https://github.com/shinyblink/Ndless/actions/runs/196901972

=> about 4:45 minutes

The previous (duplicating) approach wins by a minute, and with much smaller artifacts.

I'll revert to the tar.gz method.

**Comment by adriweb** [comment]
Any update?

I've asked @alberthdev to check if he can help finish this, too, it would be great to have this CI :)

File: .github/workflows/main.yml
Comment (fridtjof) [reply .github/workflows/main.yml:36]: What's stored in the cache is separate from artifacts.

I intended the tar.gz upload to be kind of a "nightly" toolchain build, but I guess releases are the better tool for the job. Github Actions don't let you reuse past build's artifacts, because they're not really intended to be used like that. That's why the packaged tar.gz toolchain is also cached next to the unextracted toolchain, so it just has to be uploaded again.

I think I will remove the toolchain artifact upload though, because it does take a long time. Maybe I'll come back around to implement nightly builds in some way, but I'm not quite sure how that will work with Actions.

**Comment by fridtjof** [comment]
Thanks for reminding me, @adriweb :)

Should be good now - CI times without the toolchain are down to 60-90 seconds. I'd say that is acceptable :)

When this is merged (and when I get around to it), I'm going to take a shot at macOS CI and put out another PR with that.

**Comment by fridtjof** [comment]
rebased on top of r2016 (also master as of now), CI in progress here: https://github.com/shinyblink/Ndless/actions/runs/290422251

**Comment by alberthdev** [comment]
Was asked to take a look at this - changes as they stand today look pretty good! You're also protected from [this issue](https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/), of which I'll need to update sooner rather than later as well for another project...

Some comments addressing stuff mentioned above:

 * **Security** - it might help to filter on the event's activity types, at least from the perspective of only allowing activity types that a maintainer can actually trigger, e.g. labeling "ready-for-ci" or similar. See: [pull_request event](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request)
 * **Other platforms** - if the resulting GitHub Actions is different enough that it can't really be easily done in one file, it's not a bad idea to make those distinct. Example would be if we're trying to build on Windows - that could maybe be done in this one (and only use WSL to build), or in a separate file with a build matrix toggling on or off the WSL build step. No strong opinion here though...
 * **Older Linux** - probably best to leave Docker out of the mix for now, if we ever add it maybe as a separate actions workflow file too. I'd (half-jokingly) suggest that if supporting older Linux distros is of interest, running this image could be fruitful: https://github.com/pypa/manylinux - and we could sort of rely on their judgement for maintenance. (Bonus if the resulting toolchain / binaries were outputted through [actions/upload-artifact](https://github.com/actions/upload-artifact)... then "many Linux users" would be able to run as well!)

**Comment by Vogtinator** [comment]
It's looking good now! But before I merge, I have one question about GitHub actions: If someone makes a PR to master with a changed `workflows/main.yml` file, which one would the action run with which token with what permissions from which repo?

**Comment by Danke-Boi** [comment]
What language is this in? C++? C?

**Comment by Danke-Boi** [comment]
I’d like to help, but I do not know how to program, and would like to learn
more about it. What do you suggest I do first? How did you learn?

**Comment by lights0123** [comment]
@Vogtinator no secret tokens are available in PRs (although they will be available in the creator's fork). If you merge them, however, they will become available.

**Review by Vogtinator — APPROVED** [review]

**Pull request merged**
